core/src/main/java/org/infinispan/marshall/protostream/impl/SerializationContextRegistryImpl.java

```diff
@@ -0,0 +1,141 @@
+package org.infinispan.marshall.protostream.impl;
+
+import static org.infinispan.marshall.protostream.impl.SerializationContextRegistry.MarshallerType.GLOBAL;
+import static org.infinispan.marshall.protostream.impl.SerializationContextRegistry.MarshallerType.PERSISTENCE;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Consumer;
+
+import org.infinispan.configuration.global.GlobalConfiguration;
+import org.infinispan.factories.annotations.Inject;
+import org.infinispan.factories.annotations.Start;
+import org.infinispan.factories.scopes.Scope;
+import org.infinispan.factories.scopes.Scopes;
+import org.infinispan.marshall.persistence.impl.PersistenceContextInitializerImpl;
+import org.infinispan.protostream.BaseMarshaller;
+import org.infinispan.protostream.FileDescriptorSource;
+import org.infinispan.protostream.ProtobufUtil;
+import org.infinispan.protostream.SerializationContext;
+import org.infinispan.protostream.SerializationContextInitializer;
+import org.infinispan.protostream.config.Configuration;
+
+@Scope(Scopes.GLOBAL)
+public class SerializationContextRegistryImpl implements SerializationContextRegistry {
+
+   @Inject GlobalConfiguration globalConfig;
+
+   private final MarshallerContext global = new MarshallerContext();
+   private final MarshallerContext persistence = new MarshallerContext();
+
+   @Start
+   public void start() {
+      // Add user configured SCIs to both the global and persistence context
+      List<SerializationContextInitializer> scis = globalConfig.serialization().contextInitializers();
+      update(GLOBAL, ctx -> {
+         if (scis != null)
+            ctx.addContextIntializers(scis);
+
+         ctx.addContextIntializer(new PersistenceContextInitializerImpl())
+               // Register Commons util so that KeyValueWithPrevious can be used with JCache remote
+               .addContextIntializer(new org.infinispan.commons.GlobalContextInitializerImpl())
+               .update();
+      });
+
+      update(PERSISTENCE, ctx -> {
+         if (scis != null)
+            ctx.addContextIntializers(scis);
+
+         ctx.addContextIntializer(new PersistenceContextInitializerImpl())
+               .update();
+      });
+   }
+
+   @Override
+   public SerializationContext getGlobalCtx() {
+      synchronized (global) {
```

Comment:
So this is throwing red flags at me.. why is this synchronized? Do you plan on having concurrent access to the underlying `SerializationContext`? Because there is insufficient synchronization in the collections in the `SerializationContext` class itself, if so.

For example multiple threads can still invoke `SerializationContext#addContextIntializer` at the same time.

Reply:
You're right, returning `SerializationContext` is wrong. Instead `SerializationCotnextRegistry` should expose the `ImmutableSerializationContext` interface as all updates should be performed using the registry methods.

Reply:
This technically still isn't thread safe, however I didn't realize the underlying SerializationContextImpl implements ReadWriteLocks to prevent concurrent read/write operations from happening. So this should be fine.